tests/test_atom_mapping.py

```diff
@@ -953,3 +971,281 @@ def test_get_cores_and_diagnostics():
+
+    n_mols = len(mols_with_hs)
+
+    # with_h_visits = []
+    # without_h_visits = []
+    # hybrid_visits = []
+
+    # with_h_core_sizes = []
+    # without_h_core_sizes = []
+    # hybrid_core_sizes = []
+
+    TEST_ATOM_MAPPING_KWARGS = copy.deepcopy(DEFAULT_ATOM_MAPPING_KWARGS)
+
+    # useful for testing larger cutoff settings
+    # TEST_ATOM_MAPPING_KWARGS["ring_cutoff"] = 0.4
+    # TEST_ATOM_MAPPING_KWARGS["chain_cutoff"] = 0.4
+
+    for i in range(n_mols):
+        for j in range(i + 1, n_mols):
+            mol_a_with_h, mol_b_with_h = mols_with_hs[i], mols_with_hs[j]
+            mol_a_without_h, mol_b_without_h = mols_without_hs[i], mols_without_hs[j]
+
+            cores_h, diagnostics_with_h = atom_mapping.get_cores_and_diagnostics(
+                mol_a_with_h, mol_b_with_h, **TEST_ATOM_MAPPING_KWARGS
+            )
+            core_h = cores_h[0]
+
+            cores_no_h, diagnostics_no_h = atom_mapping.get_cores_and_diagnostics(
+                mol_a_without_h, mol_b_without_h, **TEST_ATOM_MAPPING_KWARGS
+            )
+            core_no_h = cores_no_h[0]
+
+            # hybrid method, use core from without H atom-mapping
+            ntom_mol_a = new_to_old_map_after_removing_hs(mol_a_with_h)
+            ntom_mol_b = new_to_old_map_after_removing_hs(mol_b_with_h)
+
+            core_a_initial = [ntom_mol_a[x] for x in core_no_h[:, 0]]
+            core_b_initial = [ntom_mol_b[x] for x in core_no_h[:, 1]]
+
+            initial_mapping = np.stack([core_a_initial, core_b_initial], axis=1)
+            MAPPING_KWARGS_WITH_MAPPING = copy.deepcopy(TEST_ATOM_MAPPING_KWARGS)
+            MAPPING_KWARGS_WITH_MAPPING["initial_mapping"] = initial_mapping
+
+            cores_hybrid, diagnostics_hybrid = atom_mapping.get_cores_and_diagnostics(
+                mol_a_with_h, mol_b_with_h, **MAPPING_KWARGS_WITH_MAPPING
+            )
+
+            core_hybrid = cores_hybrid[0]
+
+            # useful printing for debugging
+            # tnv_h = diagnostics_with_h.total_nodes_visited
+            # tnv_hybrid = diagnostics_hybrid.total_nodes_visited
+            # tnv_no_h = diagnostics_no_h.total_nodes_visited
+            # print("TNV: all_hs, no_hs, hybrid", tnv_h, tnv_no_h, tnv_no_h + tnv_hybrid)
+            # print("CORE SIZE: all_hs, no_hs, hybrid", len(core_h), len(core_no_h), len(core_hybrid))
+
+            # with_h_visits.append(tnv_h)
+            # without_h_visits.append(tnv_no_h)
+            # hybrid_visits.append(tnv_no_h + tnv_hybrid)
+
+            # with_h_core_sizes.append(len(core_h))
+            # without_h_core_sizes.append(len(core_no_h))
+            # hybrid_core_sizes.append(len(core_hybrid))
+
+            assert len(core_no_h) < len(core_h), f"Mol {i} -> {j} failed to produce larger mapping by adding hydrogens"
+            assert len(cores_no_h) < len(
+                core_hybrid
+            ), f"Mol {i} -> {j} failed to produce larger mapping by running hybrid"
```

Comment:
uncertainty: would we also expect `len(core_hybrid) == len(core_h)`?

Reply:
It does not appear to be the case, will take a look at why that is.

Reply:
Found a pretty clear case which is that if you have several terminal atoms that are not hydrogens being transformed into hydrogens, you may build an initial map using different atoms than what the hybrid based on the hydrogen-less molecule would produce.
In the picture below the top is the full hydrogens and the bottom is the hybrid. You can see that the hybrid has the flourine mapped to another carbon which prevents the retrieval of the same size core. 
![hybrid-vs-non](https://github.com/user-attachments/assets/fed33f52-7ac0-422c-9d77-50b3baf011eb)